File: world/examples/cartpole_closed_form.py
"""
CartPole Closed-Form Controller
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Heuristic PD-like controller for CartPole-v1.

Usage:
    PYTHONPATH=. python world/examples/cartpole_closed_form.py          # console only
    PYTHONPATH=. python world/examples/cartpole_closed_form.py --render # browser WebSocket
"""

import sys
import time

import world


def main():
    env = world.make("CartPole-v1")
    episodes = 10
    max_steps = 500
    render = "--render" in sys.argv

    print("=" * 50)
    print("  CartPole-v1  ·  Closed-Form (Heuristic) Controller")
    if render:
        print("  Render mode: browser (http://localhost:8080)")
    print("=" * 50)

    if render:
        print("  Waiting for browser connection (3s)...")
        time.sleep(3)

    total_steps = 0

    for ep in range(episodes):
        obs, _ = env.reset(seed=ep * 100)
        steps = 0

        for _ in range(max_steps):
            _, _, theta, theta_dot = obs
            action = 1 if theta > 0 and theta_dot > 0.01 else \
                     0 if theta > 0 else \
                     0 if theta < 0 and theta_dot < -0.01 else 1
            result = env.step(action)
            obs = result.observation
            steps += 1

            if render:
                env.render("human")
                time.sleep(0.033)

            if result.terminated or result.truncated:
                break

        total_steps += steps
        print(f"  Episode {ep + 1:>2}: {steps:>3} steps")

    print("=" * 50)
    print(f"  Average: {(total_steps / episodes):.1f} steps")
    print("=" * 50)

    env.close()

    if render:
        print("  Press Enter to stop server...")
        input()


if __name__ == "__main__":
    main()

File: world/examples/cartpole_closed_form.ts
/**
 * examples/cartpole_closed_form.ts
 * CartPole-v1 with a closed-form (heuristic) PD-like controller.
 *
 * The controller computes force direction from theta and theta_dot:
 *   if theta > 0 (falling right) → push right
 *   if theta < 0 (falling left)  → push left
 *   theta_dot adds damping to reduce overshoot.
 *
 * Usage:
 *   npx tsx world/examples/cartpole_closed_form.ts
 */
import { CartPoleEnv } from '../envs/cartpole';

async function main(): Promise<void> {
  const env = new CartPoleEnv();
  const episodes = 10;
  const maxSteps = 500;
  const render = process.argv.includes('--render');

  console.log('='.repeat(50));
  console.log('  CartPole-v1  ·  Closed-Form (Heuristic) Controller');
  if (render) console.log('  Render mode: browser (http://localhost:8080)');
  console.log('='.repeat(50));

  if (render) {
    console.log('  Waiting for browser connection (3s)...');
    await new Promise(r => setTimeout(r, 3000));
  }

  let totalSteps = 0;

  for (let ep = 0; ep < episodes; ep++) {
    const { observation } = env.reset({ seed: ep * 100 });
    let obs = observation;
    let steps = 0;

    for (let s = 0; s < maxSteps; s++) {
      const [, , theta, theta_dot] = obs;
      const action = theta > 0
        ? (theta_dot > 0.01 ? 1 : 0)
        : (theta_dot < -0.01 ? 0 : 1);

      const result = env.step(action);
      obs = result.observation as number[];
      steps++;

      if (render) {
        env.render('human');
        await new Promise(r => setTimeout(r, 33));
      }
      if (result.terminated || result.truncated) break;
    }

    totalSteps += steps;
    console.log(`  Episode ${ep + 1}: ${steps} steps`);
  }

  console.log('='.repeat(50));
  console.log(`  Average: ${(totalSteps / episodes).toFixed(1)} steps`);
  console.log('='.repeat(50));

  if (render) {
    console.log('  Press Enter to stop server...');
    process.stdin.resume();
    await new Promise<void>(resolve => process.stdin.once('data', () => resolve()));
  }
  env.close();
}

main().catch(console.error);
